source/slang/slang-ir-legalize-varying-params.cpp

```diff
@@ -1544,4 +1547,2333 @@ void depointerizeInputParams(IRFunc* entryPointFunc)
     }
 }
 
+
+// Commony entry point legalization context for Metal and WGSL.
+class LegalizeShaderEntryPointContext
+{
+public:
+    enum class LegalizeTarget
+    {
+        Metal,
+        WGSL,
+    };
+
+    LegalizeShaderEntryPointContext(IRModule* module, DiagnosticSink* sink, LegalizeTarget target)
+        : m_module(module), m_sink(sink), m_target(target)
+    {
+    }
+
+    void legalizeEntryPoints(List<EntryPointInfo>& entryPoints)
+    {
+        for (auto entryPoint : entryPoints)
+            legalizeEntryPoint(entryPoint);
+        removeSemanticLayoutsFromLegalizedStructs();
+    }
+
+private:
+    IRModule* m_module;
+    DiagnosticSink* m_sink;
+    LegalizeTarget m_target;
+    HashSet<IRStructField*> semanticInfoToRemove;
+
+    bool isTargetMetal() const { return m_target == LegalizeTarget::Metal; }
+    bool isTargetWGSL() const { return m_target == LegalizeTarget::WGSL; }
+
+    struct SystemValueInfo
+    {
+        String systemValueName;
+        SystemValueSemanticName systemValueNameEnum;
+
+        ShortList<IRType*> permittedTypes;
+        bool isUnsupported = false;
+
+        // Only used by Metal.
+        bool isSpecial = false;
+
+        SystemValueInfo()
+        {
+            // most commonly need 2
```

Comment:
This shouldn't be needed, as shortlist by default will statically allocate 16 elements on stack.